### ISSUE TITLE
Apparent duplicates in superiors/subordinates on person detail

### ISSUE THREAD

**Issue by jeancochrane** [security-force-monitor/sfm-cms#697]
When we generate the list of commanders for the superiors/subordinates tables on person detail views, we define uniqueness by the combination of rank, role, unit, start date, and end date. However, since postings are also considered unique by title, and we don't display titles in either of these tables, the tables wind up displaying apparent duplicates, as on https://whowasincommand.com/en/person/view/a3d2cb22-2514-4e39-9444-793616bcad4d/#memberships:

<img width="894" alt="Screen Shot 2020-09-01 at 9 41 04 AM" src="https://user-images.githubusercontent.com/14170650/91865704-47ca0600-ec37-11ea-969c-e36f2b8617ae.png">

We should update the query in `person.views.get_commanders()` to group these results by title:

https://github.com/security-force-monitor/sfm-cms/blob/99c0454c23486225ca2d213dda236ab50a3e332d/person/views.py#L163-L185

I think the move here will be to adjust the `DISTINCT` call to group by person, unit, rank, role, start date, and end date. We'll also have to adjust the query to grab some of that info too, and we'll want to refactor the code that follows to use the person/unit/rank/role/dates that we retrieve from the query instead of taking an ID returned by the query and rehydrating the objects using the Django ORM.

It would be nice to practice test-driven development here and start with by speccing out tests for `get_commanders()`, one for each relationship type (`child` and `parent`), that take in a posting that has superiors and subdorindates that would appear to be dupes under the current code (i.e. the same person, unit, rank, role, and dates, just with a different title for each posting). Defining the fixture for these tests might be a little tricky if you're not comfortable with the data model so I'd be happy to help consult.

**Comment by jeancochrane** [comment]
Closed in #698.

**Comment by tlongers** [comment]
I noticed this issue is still [manifesting on some records](https://whowasincommand.com/en/person/view/58a9adff-a833-46f7-aaaf-0133aaed9284/).

**Comment by tlongers** [comment]
On our staging site this is still an issue, for example for the record of `Goodluck Ebele Jonathan`:

`~/en/person/view/bed3acf0-5d70-4e0a-b7bb-6f13ba7fc109/`

You can see the duplicates in the subordinates list:

![image](https://user-images.githubusercontent.com/2115933/108832574-9e4f4600-75c3-11eb-823c-7467161a0f4f.png)


@hancush What's your take on why this is still occuring?

**Comment by hancush** [comment]
The duplicates in the table link to the same detail page, so I don't think it's a question of the importer creating duplicates. Instead, I think it has something to do with how we build the subordinates array when we're gathering data for the detail view.

**Comment by tlongers** [comment]
Okay, I'll come up with a test case to try and unpick this, per Jean's earlier input on the issue, and then return to this issue.

**Comment by hancush** [comment]
I did a little digging and, for our most recent test case, the importer seems to be creating duplicate memberships for the person whose detail view we're looking at. We loop over these memberships to build up a list of superiors and subordinates, so we're adding those people to the array for every duplicate membership of the person detailed. (In some cases, a person can have more than one membership in a given org, so we should deduplicate this list by organization anyway, which would provide a cosmetic fix but not address the underlying data problem for affected people.)

I cannot figure out why duplicate memberships are being created for some people, but not others. One common thread seems to be that all affected people, so far, are commanders, but this could be a red herring. Will continue looking.

**Comment by hancush** [comment]
Update: I made fixes for two classes of duplicates in https://github.com/security-force-monitor/sfm-cms/compare/hcg/translation..hcg/duplicates.

1. **Multiple memberships in one organization causes duplicates in superiors and subordinates.** I addressed this by getting the distinct organizations to which a person belongs, in order to build the superior/subordinate lists.
2. **Whitespace around title, rank, and/or role causes duplicate memberships to be created.** This is tricky. In the import, we look for an existing membership based on the raw values. If we don't find it, we create a new one. When we create the new one, we do some formatting of the input data, such that the created membership has no whitespace, etc. That means, every time we import, if there is whitespace, we fail to find the membership, then create duplicates of an existing membership. I put together a band-aid for this, but there is a ton of room for improvement in this section of the import, because it's very hard to follow, opening up room for errors in data creation like this.
